src/scene/Entity.hpp

```diff
@@ -14,37 +14,68 @@
 
 #pragma once
 
+#include <utility>
+
+#include <APIObject.hpp>
+#include <RGLFields.hpp>
 #include <scene/Scene.hpp>
 #include <scene/Mesh.hpp>
-#include <APIObject.hpp>
-#include <utility>
 #include <math/Mat3x4f.hpp>
 
-
+/**
+ * Entity represents an instance of a mesh on the scene, with attributes such as transform, id, intensity texture, etc.
+ */
 struct Entity : APIObject<Entity>
 {
 	friend struct APIObject<Entity>;
 	friend struct Scene;
 
-	Entity(std::shared_ptr<Mesh> mesh, std::optional<std::string> name = std::nullopt);
-
-	// TODO(prybicki): low-prio optimization: do not rebuild whole IAS if only transform changed
-	void setTransform(Mat3x4f newTransform);
+	/**
+	 * Factory methods which creates an Entity and adds it to the given Scene.
+	 * See constructor docs for more details.
+	 */
+	static std::shared_ptr<Entity> create(std::shared_ptr<Mesh> mesh, std::shared_ptr<Scene> scene);
 
+	/**
+	 * Sets ID that will be used as a point attribute ENTITY_ID_I32 when a ray hits this entity.
+	 */
 	void setId(int newId);
-	int getId() const { return id; }
 
+	/**
+	 * Sets or updates Entity's transform.
+	 */
+	void setTransform(Mat3x4f newTransform);
+
+	/**
+	 * Sets intensity texture that will be used as a point attribute INTENSITY_F32 when a ray hits this entity.
+	 */
 	void setIntensityTexture(std::shared_ptr<Texture> texture);
 
-	OptixInstance getIAS(int idx);
+	/**
+	 * @return Estimated velocity based on current and previous transforms.
+	 */
+	Mat3x4f getVelocity() const;
 
-	std::shared_ptr<Mesh> mesh;
-	std::shared_ptr<Texture> intensityTexture = nullptr;
+	/**
+	 * @return The Scene in which this Entity is present.
+	 */
+	std::shared_ptr<Scene> getScene() const { return scene; }
```

Comment:
Technically scene returned this way is mutable, in other words, you have access to internal state of Entity. This could not matter that much (or at all), so please decide yourself, where keep const here or not.

Reply:
Modifying `Scene` does not impact `Entity` in any way. `Scene` is kept in `Entity` only as a reference, not as its part. This reference cannot be changed or removed, therefore `std::shared_ptr<Scene> getScene() const` looks like a good decision to me. WDYT?

Reply:
Fair point, you are right, this passing ptr to common scene, so no internal state is exposed. Definitely may stay const.